### DISCUSSION TITLE
Eventstore conflict retries

### DISCUSSION THREAD

**Discussion by elliotweaver** [wesone/blackrik#5] [Q&A]
In the case of an eventstore conflict I see that a conflict error is thrown. Is there a retry that occurs somewhere? I am wondering were that is or would be handled?

**Comment by wesone** [comment]
**Short answer:**
The conflict error would just cause the command to fail.

**Detailed answer:**
For example if you execute the command `doStuff` on an aggregate with the ID `a10f9ab7-2dc6-4d20-b4f2-8f79a830a66c`, the command will... well... do stuff and return an event called `STUFF_DONE`. 

But in the meantime (during the execution of the doStuff command) another command (let's say `archive`) already added the event `ARCHIVED` for the aggregate `a10f9ab7-2dc6-4d20-b4f2-8f79a830a66c`.
The command `doStuff` however is not aware of the change and now operates on an outdated state (the aggregates state before the `ARCHIVED` event). So the `STUFF_DONE` event will be rejected (conflict error) because maybe doStuff must fail if the aggregate is already archived like this:
```javascript
doStuff: async (command, state) => {
    if(state.isArchived)
        throw new BadRequestError('Aggregate already archived');

    const payload = await doSomeAsyncStuff(command.payload);
    return {
        type: 'STUFF_DONE',
        payload
    };
},
archive: async (command, state) => {
    if(state.isArchived)
        throw new BadRequestError('Aggregate already archived');

    return {
        type: 'ARCHIVED'
    };
}
```
The conflict error now makes sure that there can't be a `STUFF_DONE` event after the `ARCHIVED` event.

**Reply by elliotweaver** [reply]
Thanks for your response. I really appreciate the detailed explanation.

The use case I am thinking of is slightly different. I am thinking more of a highly concurrent single handler. For example, suppose payments occurred in an upstream service at a high throughput and events are sent to a target service for writing to a general ledger (say through a Kafka event bus). In this case, the events must be processed and the balances must be written.

If a single balance aggregate handles two such payments for a single account at the same time, one of these will fail due to the optimistic concurrency that you describe. My question is, in the ideal architecture of Blackrik, where would you envision the retry of the failed command occurring? Would you push that decision to the Kafka command handler?

**Reply by wesone** [reply]
I'd say the framework itself is not allowed to make the assumption if it's safe to retry a failed command (without correcting its payload or something). So you would build a command retry mechanism yourself. May I ask what executes the commands? Is it another server? If so, that server should resend commands in case of an error. 

If a bunch of users execute the commands (via some frontend), is it really necessary that these users operate on the same aggregate?

You wrote 
> events are sent to a target service [...] In this case, the events must be processed and the balances must be written

At that point you already have an event so the command succeeded and you have the correct system state inside the event store. Then the event bus is responsible for making sure that your ledger successfully handles the event. So if there is a temporary error inside your ledger the event bus should resend the event until the ledger processed it without error (for this to work it is important for the event handler to be idempotent).

Then you asked
> Would you push that decision to the Kafka command handler?

This also depends on how commands are executed. If there is another message broker (besides the event bus which technically may be the same Kafka instance) and you write these payments into that message broker, have a message handler on the other side that receives each message and executes a command accordingly, then you can make use of the brokers retry mechanism (if there is any) and let the message handler fail if the command failed. However this may be a little bit of an overkill.

In every case the command retry mechanism would be some system in front of the commands.
